cvat/apps/engine/cache.py

```diff
@@ -1,349 +1,700 @@
+            )
+        )
+
+    def _read_raw_images(
+        self,
+        db_task: models.Task,
+        frame_ids: Sequence[int],
+        *,
+        manifest_path: str,
+    ):
+        db_data = db_task.data
+
+        if os.path.isfile(manifest_path) and db_data.storage == models.StorageChoice.CLOUD_STORAGE:
+            reader = ImageReaderWithManifest(manifest_path)
+            with ExitStack() as es:
+                db_cloud_storage = db_data.cloud_storage
+                assert db_cloud_storage, "Cloud storage instance was deleted"
+                credentials = Credentials()
+                credentials.convert_from_db(
+                    {
+                        "type": db_cloud_storage.credentials_type,
+                        "value": db_cloud_storage.credentials,
                     }
-                    cloud_storage_instance = get_cloud_storage_instance(cloud_provider=db_cloud_storage.provider_type, **details)
+                )
+                details = {
+                    "resource": db_cloud_storage.resource,
+                    "credentials": credentials,
+                    "specific_attributes": db_cloud_storage.get_specific_attributes(),
+                }
+                cloud_storage_instance = get_cloud_storage_instance(
+                    cloud_provider=db_cloud_storage.provider_type, **details
+                )
+
+                tmp_dir = es.enter_context(tempfile.TemporaryDirectory(prefix="cvat"))
+                files_to_download = []
+                checksums = []
+                media = []
+                for item in reader.iterate_frames(frame_ids):
+                    file_name = f"{item['name']}{item['extension']}"
+                    fs_filename = os.path.join(tmp_dir, file_name)
+
+                    files_to_download.append(file_name)
+                    checksums.append(item.get("checksum", None))
+                    media.append((fs_filename, fs_filename, None))
+
+                cloud_storage_instance.bulk_download_to_dir(
+                    files=files_to_download, upload_dir=tmp_dir
+                )
+                media = preload_images(media)
+
+                for checksum, (_, fs_filename, _) in zip(checksums, media):
+                    if checksum and not md5_hash(fs_filename) == checksum:
+                        slogger.cloud_storage[db_cloud_storage.id].warning(
+                            "Hash sums of files {} do not match".format(file_name)
+                        )
+
+                yield from media
+        else:
+            requested_frame_iter = iter(frame_ids)
+            next_requested_frame_id = next(requested_frame_iter, None)
+            if next_requested_frame_id is None:
+                return
+
+            # TODO: find a way to use prefetched results, if provided
+            db_images = (
+                db_data.images.order_by("frame")
+                .filter(frame__gte=frame_ids[0], frame__lte=frame_ids[-1])
```

Comment:
Probably we should set chunk_size max limit and use here `frame__in` (in that case we do not need to iterate over `requested_frame_iter` and compare a frame with `next_requested_frame_id`). What do you think? (can be changed in a separate PR)

Reply:
I don't really like using `__in`, as it can result in very big queries, which are processed slowly. And it is possible here, as `chunk_size` is only limited by the task size now. I think it can be done once we introduce a limit on the `chunk_size`, which is planned to be done in another PR.

Reply:
> I don't really like using __in, as it can result in very big queries, which are processed slowly.

I agree, but I also said that `frame__in` should be used only when we add a `chunk_size` limit.

Reply:
Ok, I thought you're speaking about order by + limit in the resulting sql.